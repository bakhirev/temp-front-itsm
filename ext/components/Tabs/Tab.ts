import styled, { css } from 'styled-components';

import { DEFAULT_THEME } from '../common';

import { TabsSize, TAB_HEIGHT_SMALL, TAB_HEIGHT_BIG } from './constants';

interface ITabProps {
  active?: boolean;
  disabled?: boolean;
  size: TabsSize;
}

export const Tab = styled.div<ITabProps>`
  display: flex;
  align-items: center;
  transition: all 0.2s;
  user-select: none;
  &:focus,
  & > div:focus {
    outline: none;
  }

  color: ${({ active, disabled, theme }) => {
    if (disabled) {
      return active ? theme.tabs.activeTextColor : theme.color.neutral[30];
    }
    return active ? theme.tabs.activeTextColor : theme.color.neutral[50];
  }};

  ${({ disabled, theme, size }) => css`
    height: ${size === 'small' ? TAB_HEIGHT_SMALL : TAB_HEIGHT_BIG};
    cursor: ${disabled ? 'default' : 'pointer'};
    & svg {
      fill: ${disabled ? theme.color.neutral[30] : theme.color.neutral[50]};
    }
    &:focus {
      ${theme.tabs.focus}
    }
  `}

  ${({ theme, disabled }) =>
    !disabled &&
    css`
      &:hover {
        ${theme.tabs.hover}
      }
      &:active {
        ${theme.tabs.active}
      }
    `}
`;

Tab.defaultProps = {
  theme: DEFAULT_THEME,
};
