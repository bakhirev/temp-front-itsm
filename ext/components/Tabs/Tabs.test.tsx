import React from 'react';
import { shallow } from 'enzyme';

import { Tab } from './Tab';

import { Tabs } from './';

jest.mock('../common', () => ({
  DEFAULT_THEME: {},
  Z_INDEXES: {
    TOOLTIP: 777,
  },
}));

describe('Tabs', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  const menu = [
    { value: 4, label: 'Самара' },
    { value: 5, label: 'Липецк' },
  ];

  const tabsRequiredProps = {
    list: [
      { value: 1, label: 'Москва', active: true },
      { value: 2, label: 'Новгород' },
      { value: 3, label: 'Орел' },
    ],
    onChange: jest.fn(),
  };

  const tabsExtendedProps = {
    list: [
      { value: 1, label: 'Москва', active: true },
      // eslint-disable-next-line react/display-name
      { value: 2, label: 'Новгород', icon: () => <div /> },
      { value: 3, label: 'Орел', badge: 5 },
    ],
    onChange: jest.fn(),
  };

  const tabsDisabledProps = {
    list: [
      { value: 1, label: 'Москва', active: true },
      { value: 2, label: 'Новгород', disabled: true },
    ],
    onChange: jest.fn(),
  };

  it('should render component with required properties', () => {
    const wrapper = shallow(<Tabs {...tabsRequiredProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with small size', () => {
    const wrapper = shallow(<Tabs {...tabsRequiredProps} size={'small'} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component without underline', () => {
    const wrapper = shallow(<Tabs {...tabsRequiredProps} underline={false} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component with overflowMenu', () => {
    const wrapper = shallow(<Tabs {...tabsRequiredProps} overflowMenu={{ list: menu }} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render component in which tabs have icon and badge', () => {
    const wrapper = shallow(<Tabs {...tabsExtendedProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onChange when user clicks tab', () => {
    const wrapper = shallow(<Tabs {...tabsRequiredProps} />);
    const tabs = wrapper.find(Tab);
    tabs.at(1).simulate('click');
    const { onChange } = tabsRequiredProps;
    expect(onChange).toBeCalledTimes(1);
  });

  it('should not call onChange when user clicks disabled tab', () => {
    const wrapper = shallow(<Tabs {...tabsDisabledProps} />);
    const tabs = wrapper.find(Tab);
    tabs.at(1).simulate('click');
    const { onChange } = tabsRequiredProps;
    expect(onChange).toBeCalledTimes(0);
  });
});
