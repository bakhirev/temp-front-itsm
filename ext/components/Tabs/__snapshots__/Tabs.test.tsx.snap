// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs should render component in which tabs have icon and badge 1`] = `
<styled.div
  mobile={false}
  theme={Object {}}
  underline={true}
>
  <styled.div
    left={0}
    theme={Object {}}
    width={0}
  />
  <styled.div
    onKeyDown={[Function]}
    size="big"
    theme={Object {}}
  >
    <styled.div
      active={true}
      key="1"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Москва
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="2"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <styled.div
        theme={Object {}}
      >
        <icon
          height={24}
          width={24}
        />
      </styled.div>
      <Styled(styled.div)
        badge={false}
        icon={true}
        theme={Object {}}
      >
        Новгород
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="3"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={true}
        icon={false}
        theme={Object {}}
      >
        Орел
      </Styled(styled.div)>
      <styled.div
        theme={Object {}}
      >
        <Component
          kind="neutral-light-inactive"
          size="big"
          value={5}
        />
      </styled.div>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Tabs should render component with overflowMenu 1`] = `
<styled.div
  mobile={false}
  theme={Object {}}
  underline={true}
>
  <styled.div
    left={0}
    theme={Object {}}
    width={0}
  />
  <styled.div
    onKeyDown={[Function]}
    size="big"
    theme={Object {}}
  >
    <styled.div
      active={true}
      key="1"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Москва
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="2"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Новгород
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="3"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Орел
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
  <styled.div
    onKeyDown={[Function]}
    size="big"
    theme={Object {}}
  >
    <Component
      innerRef={
        Object {
          "current": null,
        }
      }
      list={
        Array [
          Object {
            "isDisabled": undefined,
            "isSelected": undefined,
            "label": "Самара",
            "value": 4,
          },
          Object {
            "isDisabled": undefined,
            "isSelected": undefined,
            "label": "Липецк",
            "value": 5,
          },
        ]
      }
      menu={
        Object {
          "alignment": "right",
        }
      }
      onChange={[Function]}
    />
  </styled.div>
</styled.div>
`;

exports[`Tabs should render component with required properties 1`] = `
<styled.div
  mobile={false}
  theme={Object {}}
  underline={true}
>
  <styled.div
    left={0}
    theme={Object {}}
    width={0}
  />
  <styled.div
    onKeyDown={[Function]}
    size="big"
    theme={Object {}}
  >
    <styled.div
      active={true}
      key="1"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Москва
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="2"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Новгород
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="3"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Орел
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Tabs should render component with small size 1`] = `
<styled.div
  mobile={false}
  theme={Object {}}
  underline={true}
>
  <styled.div
    left={0}
    theme={Object {}}
    width={0}
  />
  <styled.div
    onKeyDown={[Function]}
    size="small"
    theme={Object {}}
  >
    <styled.div
      active={true}
      key="1"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="small"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Москва
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="2"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="small"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Новгород
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="3"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="small"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Орел
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
</styled.div>
`;

exports[`Tabs should render component without underline 1`] = `
<styled.div
  mobile={false}
  theme={Object {}}
  underline={false}
>
  <styled.div
    left={0}
    theme={Object {}}
    width={0}
  />
  <styled.div
    onKeyDown={[Function]}
    size="big"
    theme={Object {}}
  >
    <styled.div
      active={true}
      key="1"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Москва
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="2"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Новгород
      </Styled(styled.div)>
    </styled.div>
    <styled.div
      key="3"
      onClick={[Function]}
      onKeyUp={[Function]}
      size="big"
      tabIndex={0}
      theme={Object {}}
    >
      <Styled(styled.div)
        badge={false}
        icon={false}
        theme={Object {}}
      >
        Орел
      </Styled(styled.div)>
    </styled.div>
  </styled.div>
</styled.div>
`;
